File: GameLibrary/src/cc/lib/utils/NoDupesMap.java
package cc.lib.utils;

import java.util.Collection;
import java.util.Map;
import java.util.Set;

/**
 * Created by chriscaron on 3/11/18.
 *
 * Will assert if try to add a duplicate key
 */

public class NoDupesMap<K,V> implements Map<K,V> {

    final Map<K,V> backingMap;

    public NoDupesMap(Map<K,V> map) {
        this.backingMap = map;
    }

    @Override
    public int size() {
        return backingMap.size();
    }

    @Override
    public boolean isEmpty() {
        return backingMap.isEmpty();
    }

    @Override
    public boolean containsKey(Object key) {
        return backingMap.containsKey(key);
    }

    @Override
    public boolean containsValue(Object value) {
        return backingMap.containsKey(value);
    }

    @Override
    public V get(Object key) {
        return backingMap.get(key);
    }

    @Override
    public V put(K key, V value) {
        if (backingMap.containsKey(key) && backingMap.get(key) != value)
            throw new IllegalArgumentException("Key '" + key + " is already mapped to a value");
        return backingMap.put(key, value);
    }

    @Override
    public V remove(Object key) {
        return backingMap.remove(key);
    }

    @Override
    public void putAll(Map<? extends K, ? extends V> m) {
        backingMap.putAll(m);
    }

    @Override
    public void clear() {
        backingMap.clear();
    }

    @Override
    public Set<K> keySet() {
        return backingMap.keySet();
    }

    @Override
    public Collection<V> values() {
        return backingMap.values();
    }

    @Override
    public Set<Entry<K, V>> entrySet() {
        return backingMap.entrySet();
    }

    @Override
    public String toString() {
        StringBuffer b = new StringBuffer();
        for (K key : backingMap.keySet()) {
            b.append(key).append("=").append(backingMap.get(key)).append("\n");
        }
        return b.toString();
    }
}

File: GameLibrary2/src/cc/lib/utils/NoDupesMap.kt
package cc.lib.utils

/**
 * Created by chriscaron on 3/11/18.
 *
 * Will assert if try to add a duplicate key
 */
class NoDupesMap<K, V>(val backingMap: MutableMap<K, V>) : MutableMap<K, V> {

	override val size: Int
		get() = backingMap.size

	override fun isEmpty(): Boolean {
		return backingMap.isEmpty()
	}

	override fun containsKey(key: K): Boolean {
		return backingMap.containsKey(key)
	}

	override fun containsValue(value: V): Boolean {
		return backingMap.containsValue(value)
	}

	override operator fun get(key: K): V? {
		return backingMap.get(key)
	}

	override fun put(key: K, value: V): V? {
		require(!(backingMap.containsKey(key) && backingMap[key] !== value)) { "Key '$key' is already mapped to a value" }
		return backingMap.put(key, value)
	}

	override fun remove(key: K): V? {
		return backingMap.remove(key)
	}

	override fun putAll(m: Map<out K, V>) {
		backingMap.putAll(m)
	}

	override fun clear() {
		backingMap.clear()
	}

	override val keys: MutableSet<K>
		get() = backingMap.keys

	override val values: MutableCollection<V>
		get() = backingMap.values

	override val entries: MutableSet<MutableMap.MutableEntry<K, V>>
		get() = backingMap.entries

	override fun toString(): String {
		val b = StringBuffer()
		for (key in backingMap.keys) {
			b.append(key).append("=").append(backingMap[key]).append("\n")
		}
		return b.toString()
	}
}
